src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIStructure.java

```diff
@@ -143,101 +371,179 @@ check if we have not found a block (when block == null
+     * Load the structure, special builder use with workOrders.
+     * Extracts data from workOrder and hands it to generic loading.
      */
-    private BlockPos getWorkingPosition()
+    public void loadStructure()
     {
-        return getWorkingPosition(0);
+        WorkOrderBuild workOrder = null;
+        if(job instanceof JobBuilder)
+        {
+            workOrder = ((JobBuilder) job).getWorkOrder();
+        }
+
+        if (workOrder == null)
+        {
+            return;
+        }
+
+        final BlockPos pos = workOrder.getBuildingLocation();
+        if (!(workOrder instanceof WorkOrderBuildDecoration) && worker.getColony().getBuilding(pos) == null)
+        {
+            Log.getLogger().warn("AbstractBuilding does not exist - removing build request");
+            worker.getColony().getWorkManager().removeWorkOrder(workOrder);
+            return;
+        }
+
+        int tempRotation = 0;
+        if (workOrder.getRotation() == 0 && !(workOrder instanceof WorkOrderBuildDecoration))
+        {
+            final IBlockState blockState = world.getBlockState(pos);
+            if (blockState.getBlock() instanceof AbstractBlockHut)
+            {
+                tempRotation = BlockUtils.getRotationFromFacing(blockState.getValue(AbstractBlockHut.FACING));
+            }
+        }
+        else
+        {
+            tempRotation = workOrder.getRotation();
+        }
+
+        loadStructure(workOrder.getStructureName(), tempRotation, pos);
+
+        workOrder.setCleared(false);
+        workOrder.setRequested(false);
+
+        requestMaterials();
     }
 
     /**
-     * Calculates the working position.
-     * <p>
-     * Takes a min distance from width and length.
-     * <p>
-     * Then finds the floor level at that distance and then check if it does contain two air levels.
-     *
-     * @param offset the extra distance to apply away from the building.
-     * @return BlockPos position to work from.
+     * Searches a handy block to substitute a non-solid space which should be guaranteed solid.
+     * @param location the location the block should be at.
+     * @return the Block.
      */
-    private BlockPos getWorkingPosition(final int offset)
+    public abstract Block getSolidSubstitution(BlockPos location);
+
+    /**
+     * Loads the structure given the name, rotation and position.
+     * @param name the name to retrieve  it.
+     * @param rotateTimes number of times to rotate it.
+     * @param position the position to set it.
+     */
+    public void loadStructure(@NotNull final String name, int rotateTimes, BlockPos position)
     {
-        if (offset > MAX_ADDITIONAL_RANGE_TO_BUILD)
+        if(job instanceof AbstractJobStructure)
         {
-            return currentStructure.getCurrentBlockPosition();
+            rotation = rotateTimes;
+            try
+            {
+                 StructureWrapper wrapper = new StructureWrapper(world, name);
```

Comment:
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) Local variable 'wrapper' could be declared final [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://home.kk-sc.de:9000/coding_rules#rule_key=pmd%3ALocalVariableCouldBeFinal)